### PR TITLE
Filter audio feedback

### PR DESCRIPTION
**Pull request by williamluer**
Filter TTS Voice out from audio. Uses pyannote speaker diarization and cosine similarity between embeddings to filter TTS voices.

This entire process takes 100-500 milliseconds but can be further optimized. 

Requires `HF_AUTH_TOKEN` to be set

### CODE REVIEW

File: openduck-py/openduck_py/utils/speaker_identification.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        for i, dist in enumerate(v["distances"]):
+            if dist > threshold:
+                new_v["times"].append(v["times"][i])
+                new_v["distances"].append(dist)
+        if new_v["times"]:
+            new_d[k] = new_v
+    return new_d
+
+
+import numpy as np
```
Comment (sjkoelle) [inline openduck-py/openduck_py/utils/speaker_identification.py:102]: move to top

**Comment by sjkoelle** [comment]
should we wait to see if this approach or webrtc is preferable prior to merging?

File: openduck-py/openduck_py/utils/speaker_identification.py (hunk 1 above)
Comment (sjkoelle) [reply openduck-py/openduck_py/utils/speaker_identification.py:102]: maybe we should use an automated linter - we have a codeant subscription

File: openduck-py/openduck_py/routers/voice.py [hunk 2]
```diff
@@ -133,6 +145,17 @@ def _inference(sentence: str):
 
         loop = asyncio.get_running_loop()
 
+        rec_time = time()
+        # wavfile.write(f"{rec_time}_before.wav", 16000, audio_data)
```
Comment (sjkoelle) [inline openduck-py/openduck_py/routers/voice.py:149]: remove this and other commends

**Comment by williamluer** [comment]
> should we wait to see if this approach or webrtc is preferable prior to merging?

No I think we should just delete it if the webrtc is better. This is a good baseline that actually enables conversations

File: openduck-py/openduck_py/utils/speaker_identification.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+
+HF_AUTH_TOKEN = os.getenv("HF_AUTH_TOKEN")
+
+
+def load_pipelines() -> tuple:
+    """
+    Load the diarization and embedding pipelines.
+    """
+    pipeline = Pipeline.from_pretrained(
```
Comment (sjkoelle) [inline openduck-py/openduck_py/utils/speaker_identification.py:19]: maybe add note that this uses the HF cache (assuming thats how this model is cached)?

**Review by sjkoelle — APPROVED** [review]

File: openduck-py/openduck_py/utils/speaker_identification.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+from typing import Dict, List
+
+import torch
+import numpy as np
+from scipy.spatial.distance import cdist
+from pyannote.core import Segment
+from pyannote.audio import Pipeline, Inference, Model
+
+
+HF_AUTH_TOKEN = os.getenv("HF_AUTH_TOKEN")
```
Comment (zachwe) [inline openduck-py/openduck_py/utils/speaker_identification.py:12]: Please add in `.env.example`

Comment (williamluer) [reply openduck-py/openduck_py/utils/speaker_identification.py:12]: added

File: openduck-py/openduck_py/utils/speaker_identification.py (hunk 3 above)
Comment (williamluer) [reply openduck-py/openduck_py/utils/speaker_identification.py:19]: What do you think this note should say and why is it important to leave as a note?

**Pull request merged**